junebug/logging_service.py

```diff
@@ -122,3 +123,61 @@ def stopService(self):
 
     def registered(self):
         return self.log_observer in self.logger.observers
+
+
+def reverse_read(filename, buf):
+    '''
+    Read the non-blank lines from a file in reverse order.
+
+    :param str filename: The path of the file to be read.
+
+    :returns:
+        A generator that yields each line from the file in reverse order as a
+        string.
+    '''
+    with open(filename) as f:
+        f.seek(0, os.SEEK_END)
+        total_size = remaining_size = f.tell()
+        offset = 0
+        incomplete_line = None
+        while remaining_size > 0:
+            offset = min(total_size, offset + buf)
+            f.seek(-offset, os.SEEK_END)
+            data = f.read(min(remaining_size, buf))
+            remaining_size -= buf
```

Comment:
This should be `-= len(data)` since the number of bytes requested might have been `remaining_size` and `.read()` may return fewer bytes than requested.